### ISSUE TITLE
Regenerate contentful/export.json

### ISSUE THREAD

**Issue by wKovacs64** [coreybruyere/coreybruyere-v3#13]
Hey @coreybruyere. I saw the mention in the Gatsby issue about CSS-in-JS theming, but it's been so long, I don't remember all the details of what was going on at the time I left those comments. I thought I might check out your site to see if I could help, but it looks like you've changed the content models a bit since they were originally exported from Contentful (added "Work", "HomePage", etc.). Can you regenerate `contentful/export.json` so I can start your site up locally? Thanks!

**Comment by coreybruyere** [comment]
@wKovacs64 - sorry just seeing this now for some reason. Thanks for opening this issue and attempting to help out. How would I go about regenerating that file? I can't find anything in [docs](https://github.com/contentful/contentful-export)

Since I posted in that thread I just removed theming after becoming frustrated with my inability to resolve the issue. I even decided on moving onto v4 (for more reasons than just theming) so that I could utilize theme-ui since I get easier theming out of the box.

**But**, I would like to still like to resolve this issue on this repo here since I won't be getting to v4 of my site for awhile. Once I figure out how to regenerate that JSON dump I'll ping you here. 

Thanks again for the help.

**Comment by coreybruyere** [comment]
And now I'm getting issues trying to build and serve 😠

**Comment by wKovacs64** [comment]
No worries. As for how to export, how did you get that file initially? Through their web interface, perhaps? It's been a while, but I think I manually did it with their CLI tool. [[`docs`](https://www.contentful.com/developers/docs/tutorials/cli/import-and-export/)]

I've moved to Theme UI for a few of my projects as well. It's not quite as nice as "manual" theming (for users) if you have to support IE11 (you'll get that "flash" on *all* browsers) like I do, but if you don't, it's great and way easier.

Another approach is the combination of [`use-dark-mode`](https://github.com/donavon/use-dark-mode#readme) and [`gatsby-plugin-use-dark-mode`](https://github.com/wKovacs64/gatsby-plugin-use-dark-mode#readme).

> And now I'm getting issues trying to build and serve

Frustrating, but I'm sure you'll get it. 🙂

**Comment by coreybruyere** [comment]
Thanks. So I updated and regenerated `contentful/export.json` and pushed to my development branch. 

I was hoping to resolve using the `useLayoutEffect`(?)  method you had noted in that Gatsby thread. I'm using rebass and have all my themes defined in an theme object so using a CSS class appended to the body (use-dark-mode route) won't really work in my case.

**Comment by wKovacs64** [comment]
By the way, `NODE_ENV` is `production` at build time, but `undefined` at runtime. That's why you're getting that issue with trying to build and serve. You can either change the `dotenv` logic to default to a production .env file or you can install `cross-env` and explicitly set `NODE_ENV=production` in your `serve` command:

```diff
-    "serve": "npm run build && gatsby serve",
+    "serve": "npm run build && cross-env NODE_ENV=production gatsby serve",
```

Either approach should work.

**Comment by wKovacs64** [comment]
So I just uncommented all the theme toggle stuff in `Menu.jsx` and it seems to work fine as far as I can tell. You're not experiencing the hydration problem that an effect would work around. Is there some particular behavior you're trying to solve or add?

**Comment by coreybruyere** [comment]
Thanks for pointing out my env issue! I'll need to look into `cross-env` and how that is resolving the issue I was having. I followed the contentful-gatsby starter docs when I initially set this up, so I'm not all entirely sure why my `NODE_ENV` is undefined at runtime?


As for theme toggle issues, the issue is only visible on production builds. To reproduce visit my branch `fixNodeEnv` :

1. run `npm run serve`
2. Toggle theme to dark mode
3. Hard refresh your browser (CMD + SHIFT + R)

After those steps you should notice that all components that reference my theme (in this case on the home page it's the gradient in my Work section, and the cards that display posts) reference the previous, light theme value after hard refreshing (instead of persisting the dark value theme state).

I created a GIF to visualize what I'm seeing. The second change in colors is where I'm doing a hard refresh.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/3430222/83664833-67da2280-a57f-11ea-9408-041b9bfefba0.gif)

Thanks again for the help. Let me know if I could buy you a coffee or beer.

**Comment by wKovacs64** [comment]
> I'll need to look into `cross-env` and how that is resolving the issue I was having.

Don't worry too much about `cross-env` itself, that's just a cross-platform (Windows, macOS, Linux) way to set an environment variable. The solution is the act of setting the environment variable (`NODE_ENV=production`) when you run `gatsby serve`. You could ditch `cross-env` and just do `NODE_ENV=production gatsby serve` - it just wouldn't work on Windows, so I always use `cross-env` because... why not? :) 

As mentioned, an alternate solution would be to change the `dotenv` logic in `gatsby-config.js` to fall back to a production env file in the event `NODE_ENV` is not set, if that makes more sense to you. Something like this:

```
require('dotenv').config({
  path: `.env.${process.env.NODE_ENV || 'production'}`,
})
```

That all said, I think you've discovered a bug in Gatsby. Their [docs](https://www.gatsbyjs.org/docs/environment-variables/) say `NODE_ENV` should be `production` during both `gatsby build` and `gatsby serve` but I just tested it in a new project and it's `undefined` in `gatsby serve` (which is what you're running into and why manually setting it helps). You should log an issue (and possibly submit a fix and get some [free swag](https://www.gatsbyjs.org/contributing/contributor-swag/)!). That would be cool. 😎 

> As for theme toggle issues, the issue is only visible on production builds. To reproduce...

Ahh okay, I see it now. I was doing it too quickly and only paid attention to the background (which stays dark). I'll try to investigate when I have some more time.

> Let me know if I could buy you a coffee or beer.

No way, just happy to help. 👍

**Comment by wKovacs64** [comment]
Ultimately, the problem you are facing is a React rehydration issue with regards to statically generated sites. This blog post does a pretty good job of explaining it (with several links to others complaining about the issue):

https://blog.logrocket.com/fixing-gatsbys-rehydration-issue/

(Note: the article has a few minor mistakes, e.g. calling it "server side rendering" when Gatsby sites are actually statically generated, but the overall point is on target.)

They explicitly go over a `useEffect` solution similar to what I alluded to in my original comment in that Gatsby thread. I took the liberty of refactoring your site to use a version of it [here](https://github.com/wKovacs64/coreybruyere-v3/commit/717b9eba88131c73184bddd1c8a41640b34fa463) if you want to take a peek. The downside to this approach is everything below the `ThemeProvider` (which is essentially everything) gets re-rendered when the client finishes booting up React, and you also see a flash-of-default-themed-content if you are using the non-default theme (dark mode, in this case). Two ways around the flash are CSS custom properties or CSS classes that don't rely on CSS-in-JS to resolve theme values, but you're not doing either of those things in this version of the site so the flash is a necessary evil.

> I'm using rebass and have all my themes defined in an theme object so using a CSS class appended to the body (use-dark-mode route) won't really work in my case.

You can still take advantage of something like `use-dark-mode` (if you want). It happens to set a class, but you can ignore that and just rely on it for tracking your dark mode state and passing the appropriate theme object to Emotion's `ThemeProvider` (per the [Rebass theming docs](https://rebassjs.org/theming/)). I used it in the commit linked above and it allowed me to delete all your custom context stuff. Just one way to go, though. YMMV.

**Comment by coreybruyere** [comment]
Thank you again for all of this. This is a solid temporary solution until I get around to rewriting a simplified v4 (no real need for a CMS, want to use theme-ui [to get custom properties out of the box for free] , and using TS next time around). 

I'm content with this solution for the time being. Spelling out the issue at hand and pointing to me to the right resources really helped me to understand what was going on. I was just about to scrap it all and start over before I cooled down and told myself I could temporarily do without a dark theme. 😆 

I'll definitely create a ticket for that Gatsby bug you noted above. However, I'm pretty sure I encountered similar threads when I was in the weeds with this issue last month. Either way, I'll investigate and replicate it and create a ticket if it's still not resolved.

Thanks again!
